### PR TITLE
Resolve join/leave race conditions from the client. Closes #1485

### LINKED ISSUE #1485: Store clients join ref and relay it from the transport when pushing messages
reference: https://github.com/phoenixframework/phoenix/issues/1483

### PR DESCRIPTION
**Pull request by chrismccord**
- Transport store join ref with channel pid and relay when channel closes/errors.
- Transport closes existing channel if client races a duplicate join.
- Last join wins and client properly filters channels specific phx_close/phx_error events.

@josevalim if you have time I could use a review here since it will effect @sasa1977's recent work.

### CODE REVIEW

**Comment by chrismccord** [comment]
@jfrolich can you give this PR a shot and see if it accommodates your previous issues?

**Comment by imranismail** [comment]
I'll try this one out tomorrow! Having the same issue as @jfrolich

**Comment by jfrolich** [comment]
Awesome will try this out tomorrow, @chrismccord (sorry have to tune my notifications, didn't see this)

File: lib/phoenix/socket/transport.ex [hunk 1]
```diff
@@ -237,6 +237,13 @@ defmodule Phoenix.Socket.Transport do
     end
   end
 
+  defp do_dispatch(pid, %{event: "phx_join"} = msg, socket) when is_pid(pid) do
+    Logger.debug "Duplicate channel join for topic \"#{msg.topic}\" in #{inspect(socket.handler)}. " <>
```
Comment (josevalim) [inline lib/phoenix/socket/transport.ex]: Should be warn per our discussion yesterday.

**Comment by josevalim** [comment]
@chrismccord it looks good but I feel like we should document this behaviour somewhere. It cannot be buried inside Phoenix.Channel.Server or such though. It needs to be facing the user API. Ideas?

**Comment by chrismccord** [comment]
@josevalim perhaps a note in Phoenix.Transport as well as a comment in phoenix.js under a "Duplicate Subscriptions" heading?

**Comment by josevalim** [comment]
Sounds good, Your call!

On Wednesday, March 30, 2016, Chris McCord <notifications@github.com> wrote:

> @josevalim <https://github.com/josevalim> perhaps a note in
> Phoenix.Transport as well as a comment in phoenix.js under a "Duplicate
> Subscriptions" heading?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/pull/1615#issuecomment-203566435>
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by chrismccord** [comment]
@imranismail @jfrolich have you had a chance to try this out?

**Comment by jfrolich** [comment]
I'm sorry, I can't reproduce my problem in a real world app easily anymore because I switched approaches (using subscriptions within channels). I might be able to write a test case for this later, but for now this looks good!

**Comment by chrismccord** [comment]
@jfrolich can you checkout a previous commit and retry this quickly for me? It would be a big help. Thanks!

**Comment by jfrolich** [comment]
Ok, let me try and see if I manage to get that working :)

**Comment by jfrolich** [comment]
Ok. I manage to recover the old state of my app. And brought in this branch of phoenix + phoenix.js. I couldn't trigger the rejoin bug in the new branch, so at least the problem I had before seems to have been fixed with this! Thanks!

Cheers,

Jaap

**Comment by chrismccord** [comment]
Thank you for verifying!

**Pull request merged**